core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java

```diff
@@ -2880,11 +2889,74 @@ private void registerQuery(
       }
       break;
 
+    case CREATE_PROCEDURE:
+      SqlCreateProcedure createProcedure = (SqlCreateProcedure) node;
+      nodeToTypeMap.put(createProcedure, getUnknownType());
+      registerQuery(parentScope, null, createProcedure.statement,
+          node, null, false, false);
+      break;
+
+    case BEGIN_END:
+    case FOR_STATEMENT:
+    case WHILE_STATEMENT:
+    case REPEAT_STATEMENT:
+    case LOOP_STATEMENT:
+      SqlLabeledBlock labeledBlock = (SqlLabeledBlock) node;
+      BlockScope blockScope = new BlockScope(parentScope, labeledBlock);
+      scopes.put(labeledBlock, blockScope);
+      LabeledBlockNamespace labeledBlockNs =
+          new LabeledBlockNamespace(this, labeledBlock, enclosingNode);
+      String blockAlias = deriveAlias(labeledBlock, nextGeneratedId++);
+      registerNamespace(blockScope, blockAlias, labeledBlockNs, false);
+      registerStatementList(blockScope, blockScope, blockAlias,
+          labeledBlock.statements, labeledBlock);
+      break;
+
+    case IF_STATEMENT:
+      SqlIfStmt ifStmt = (SqlIfStmt) node;
+      registerStatementList(parentScope, usingScope, alias,
+          ifStmt.conditionalStmtListPairs, ifStmt);
+      registerStatementList(parentScope, usingScope, alias,
+          ifStmt.elseStmtList, ifStmt);
+      break;
+
+    case CASE_STATEMENT:
+      SqlCaseStmt caseStmt = (SqlCaseStmt) node;
+      registerStatementList(parentScope, usingScope, alias,
```

Comment:
It seems like IF and CASE share a lot of duplicate code. Suggest creating a base class (e.g. SqlConditional) that both derive from. And then implement all of the shared logic in one place.

Reply:
That's a good idea, I'll update it to move some logic out of IF and CASE.